crates/services/upgradable-executor/src/executor.rs

```diff
@@ -0,0 +1,281 @@
+/// (if the database has a corresponding bytecode).
+pub struct Executor<S, R> {
+    pub storage_view_provider: S,
+    pub relayer_view_provider: R,
+    pub config: Arc<Config>,
+    #[cfg(feature = "wasm-executor")]
+    engine: wasmtime::Engine,
+    #[cfg(feature = "wasm-executor")]
+    module: wasmtime::Module,
+}
+
+impl<S, R> Executor<S, R> {
+    pub fn new(
+        storage_view_provider: S,
+        relayer_view_provider: R,
+        config: Arc<Config>,
+    ) -> Self {
+        Self {
+            storage_view_provider,
+            relayer_view_provider,
+            config,
+            #[cfg(feature = "wasm-executor")]
+            engine: crate::DEFAULT_ENGINE.clone(),
+            #[cfg(feature = "wasm-executor")]
+            module: crate::COMPILED_UNDERLYING_EXECUTOR.clone(),
+        }
+    }
+}
+
+impl<D, R> Executor<D, R>
+where
+    R: AtomicView<Height = DaBlockHeight>,
+    R::View: RelayerPort + Send + Sync + 'static,
+    D: AtomicView<Height = BlockHeight> + Modifiable,
+    D::View: KeyValueInspect<Column = Column> + Send + Sync + 'static,
+{
+    #[cfg(any(test, feature = "test-helpers"))]
+    /// Executes the block and commits the result of the execution into the inner `Database`.
+    pub fn execute_and_commit(
+        &mut self,
+        block: fuel_core_types::services::executor::ExecutionBlock,
+    ) -> fuel_core_types::services::executor::Result<ExecutionResult> {
+        let (result, changes) = self.execute_without_commit(block)?.into();
+
+        self.storage_view_provider.commit_changes(changes)?;
+        Ok(result)
+    }
+}
+
+impl<S, R> Executor<S, R>
+where
+    S: AtomicView<Height = BlockHeight>,
+    S::View: KeyValueInspect<Column = Column> + Send + Sync + 'static,
+    R: AtomicView<Height = DaBlockHeight>,
+    R::View: RelayerPort + Send + Sync + 'static,
+{
+    #[cfg(any(test, feature = "test-helpers"))]
+    /// Executes the block and returns the result of the execution with storage changes.
+    pub fn execute_without_commit(
+        &self,
+        block: fuel_core_types::services::executor::ExecutionBlock,
+    ) -> fuel_core_types::services::executor::Result<UncommittedResult<Changes>> {
+        self.execute_without_commit_with_coinbase(block, Default::default(), 0)
+    }
+
+    #[cfg(any(test, feature = "test-helpers"))]
+    /// The analog of the [`Self::execute_without_commit`] method,
+    /// but with the ability to specify the coinbase recipient and the gas price.
+    pub fn execute_without_commit_with_coinbase(
+        &self,
+        block: fuel_core_types::services::executor::ExecutionBlock,
+        coinbase_recipient: fuel_core_types::fuel_types::ContractId,
+        gas_price: u64,
+    ) -> fuel_core_types::services::executor::Result<UncommittedResult<Changes>> {
+        let component = match block {
+            ExecutionTypes::DryRun(_) => {
+                panic!("It is not possible to commit the dry run result");
```

Comment:
I know this is just a copy and paste, but should we return an error instead of a `panic` here?

Reply:
This function is marked `#[cfg(any(test, feature = "test-helpers"))]` and used only in tests, so it is okay to panic=)